active/0000-control-transpilation.md

```diff
@@ -0,0 +1,99 @@
+
+The idea is to still have addons control their own transpilation process, but
+make them honour user preferences by default.
+
+There are three different scenarios for addons:
+
+#### The addon doesn't care about the transpilation process.
+This is the most common common situation. Most addons are authored in ES6 and they don't
+mess with the host app's configuration.
+Those addons should just transpile their code user the host's apps settings.
+
+#### Addons that require experimental features
+There are two options here. Either addons can unconditionally enable that feature only for their
+tree, or it's the user responsability to enable that feature in order to use the addon, with
+optionally the addon doing some feature detection to warn (not force) the user to enable that feature.
+
+By example, an addon that uses `async-await` which right now are only available in Chrome Canary
+could either use the host app's configuration plus this feature or it could just assume the user
+has enabled it and tell them to do so in the docs.
+
+#### Addons that add their custom babel plugins
+
+The number of addons that does this is even smaller, but one of them is nothing less than Ember Data,
+which uses a babel-plugin to strip Heimdall's instrumentation from production builds.
+
+Since addons remain in control of their transpilation and just reuse the app's configuration, addons
+can attach their own plugins.
+
+#### How can addons be in charge of their own transpilation and yet honor user preferences?
+
+Ember-cli should provide to each addon a deep clone of the configuration defined in the host app
+(by example available in `this.options.babel`), and that config is the one used to transpile the addon.
+
+The addon will therefore honor the app's configuration by default, but it can also decide to add
+custom plugins, enable specific features or just throw a warning message if the app's config
+doesn't have required specific transformation enabled.
+
+By modifying this clone, their changes will not affect the host app or other addons.
+
+# How We Teach This
+
+This has no impact for apps, just for addons.
+
+Most addons will not require any change. Only those that want to modify this the transpilation
+process like [ember-computed-decorators](https://github.com/rwjblue/ember-computed-decorators/blob/master/ember-cli-build.js)
+or ember data will have to be updated.
+
+# Drawbacks
+
+We give the users more control over the transpilation process at the expenses of some complexity
+for addon authors that want to use experimental ES2017 features in their addons, leaving
+up to them to give a good user experience.
+
+# Alternatives
+
+None that I can think of, beside of not doing this.
+
+# Unresolved questions
+
+For this approach it doesn't matter if the version of babel is 5 or 6, however the plugins and their
+names are funamentally incompatible.
+
+- Should addons that depend on `ember-cli-babel` be compiled with their version or with the app's version?
```

Comment:
Im not sure.

Allowing the system to progressively transition has its benefits. For example, as add-ons upgrade more and more of the users code-gen would no longer need to cater to the lowest common denominator (ES3).

This would allow our users to see incremental progress. Especially if the most popular / largest add-ons (like ember-data or even ember) made the change more quickly.

The alternative, requiring homogenous setup. Requires the entire universe to transition before progress is seen.

I suspect coupling incremental + a Quest issue, would enable the best outcomes. (e.g. incremental rewards for our community)